packages/util/index.d.ts

```diff
@@ -33,12 +35,77 @@ declare function canPrefetch<Client, ClientOptions> (
   options: Options<Client, ClientOptions>
 ): boolean
 
-declare function getInternal (
-  variables: any,
-  request: middy.Request
-): Promise<any>
+type InternalOutput<TVariables> = TVariables extends string[] ? { [key in TVariables[number]]: unknown } : never
+
+// get an empty object if false is passed
+declare function getInternal<
+  TContext extends LambdaContext,
+  TInternal extends Record<string, unknown>
+> (
+  variables: false,
+  request: middy.Request<unknown, unknown, unknown, TContext, TInternal>
+): Promise<{}>
+
+// get all internal values if true is passed (with promises resolved)
+declare function getInternal<
+  TContext extends LambdaContext,
+  TInternal extends Record<string, unknown>
+> (
+  variables: true,
+  request: middy.Request<unknown, unknown, unknown, TContext, TInternal>
+): Promise<DeepAwaited<TInternal>>
+
+// get a single value
+declare function getInternal<
+  TContext extends LambdaContext,
+  TInternal extends Record<string, unknown>,
+  TVars extends keyof TInternal | string
+> (
+  variables: TVars,
+  request: middy.Request<unknown, unknown, unknown, TContext, TInternal>
+): TVars extends keyof TInternal
+  ? Promise<DeepAwaited<{ [_ in SanitizeKey<TVars>]: TInternal[TVars] }>>
+  : TVars extends string
+    ? IsUnknown<Choose<DeepAwaited<TInternal>, TVars>> extends true
+      ? unknown // could not find the path
+      : Promise<{ [_ in SanitizeKey<TVars>]: Choose<DeepAwaited<TInternal>, TVars> }>
+    : never // path is not a string or a keyof TInternal
```

Comment:
@lmammino  I'm gonna leave this comment just here, but what I'm saying applies everywhere `// path is not a string or a keyof TInternal` appears and also for the return type of the now limited `DeepAwaited` recursive type function.

`never` should be used as return type only if the function is gonna throw in that case, otherwise the user will get a value of type `never`, `never` is a subtype of everything and guess what 😬😂. `unknown` is more appropriate as a return type if the user call the function with invalid input and the function is not gonna throw. If you know that the function will return `undefined`/`null`/`Promise<undefined>` in that case, use that type.

P.S. Just to let you know, `Promise<never>` should be used only for a promise that will never be fulfilled, or that will get rejected for sure.

Reply:
Thanks, very good point! Resolved now